ObjectCreation/AeroFoil.FCMacro

```diff
@@ -0,0 +1,1747 @@
+### 		AEROFOIL MACRO CALLS - Top (check Bottom)		###
+###---------------------------------------------------------------------###
+									###
+def AeroFoil_generate(obj):						###
+    n_ = (obj.multi * (obj.multiParam - 1)) + 1				###
+    AeroFoil_object = AeroFoil(obj)					###
+    if AeroFoil_object.create():					###
+        AeroFoil_object.characterize()					###
+    else:								###
+        return False							###
+    AeroFoil_object.copy(n_)						###
+    return True								###
+									###
+###---------------------------------------------------------------------###
+### 		AEROFOIL MACRO CALLS - Top (check Bottom)		###
+###---------------------------------------------------------------------###
+###########################################################################
+
+
+
+# Constant Variables
+# ------------------------------------------------------------------------------------------------
+
+MAX_REFINE_PARAM = 2
+MAX_QUANTITY_PARAM = 5
+MIN_DATA_POINTS = 10
+NACA_NUMBER_OF_POINTS = 50
+NACA_5_2ND_3RD_DIGITS = ["10", "20", "30", "40", "50", "21", "31", "41", "51"]
+FUNCTIONSARRAY = [
+    "+",
+    "-",
+    "*",
+    "/",
+    "^",
+    "e",
+    "pi",
+    "ln(",
+    "log(",
+    "sqrt(",
+    "sin(",
+    "cos(",
+    "tan(",
+    "asin(",
+    "acos(",
+    "atan(",
+    "x",
+    "(",
+    ")",
+    ".",
+]
+UNITSCONVERSION = [0, 1, 10, 1000, 25.4, 304.8, 914.4]
+MIN_CHORD_LENGTH_MM = 1
+MD_ = App.getUserMacroDir(True) + "/AeroFoil_UI_Files/AeroFoil_"
+
+
+
+# Global Variables
+# ------------------------------------------------------------------------------------------------
+
+# Variables Pertaining to the Entire Program
+dialogIndex = [1, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11]
+
+
+
+# Main Function Class
+# ------------------------------------------------------------------------------------------------
+
+class AeroFoil:
+    def __init__(self, obj):
+        global UNITSCONVERSION
+        self.profileName = generateName("AeroFoil")
+        self.chord = obj.chordLength * UNITSCONVERSION[obj.chordUnits]
+        self.designtype = [obj.workbench, obj.curveType]
+        self.designclosed = obj.closed_
+        self.points = []
+        self.points.append(obj.pointsX)
+        self.points.append(obj.pointsY)
+        self.n_units = len(self.points[0])
+        self.objRefData = [
+            obj.airfoilType,
+            obj.airfoil4DNumber,
+            obj.airfoil5DNumber,
+            obj.airfoilProfileType,
+            obj.importFrom,
+        ]
+        self.designprogressbar = obj.progressBar_
+
+    def copy(self, quantity_):
+        if quantity_ != 1:
+            refNum = int(re.findall(r"\d+", self.profileName)[0])
+            FreeCADGui.Selection.clearSelection()
+            FreeCADGui.Selection.addSelection(
+                App.ActiveDocument.getObjectsByLabel(self.profileName)[0]
+            )
+            FreeCADGui.runCommand("Std_Copy")
```

Comment:
I'm surprised that you need to use copy-paste. Can't you duplicate what you want to duplicate directly in Python? This copy-paste thing appears to be to be more a hack and I have doubts about the robustness of the method. I can help you with pleasure with that matter.

Reply:
Using simple copy did cross my mind; but it creates an object. For the sketches and dwires, I'd prefer to retain their editable characteristics. Hence, the copy and paste. 
Duplicate, I think, just does it once. And then you have do a addSelection method again, and again. 
With the copy-paste method, I copy once, and then paste N times. Also, as the macro is modal, the user cannot do an external 'copy' while the macro is running.

Reply:
Thank you Mr. Galou. Using simple copy did cross my mind; but it creates an object. For the sketches and dwires, I'd prefer to retain their editable characteristics. Hence, the copy and paste.
Duplicate, I think, just does it once. And then you have do a addSelection method again, and again.
With the copy-paste method, I copy once, and then paste N times. Also, as the macro is modal, the user cannot do an external 'copy' while the macro is running.

Reply:
I just found a **safe alternative** to the former copy-paste hack method on the [Forum](https://forum.freecadweb.org/viewtopic.php?t=11014).
It does just that (i.e., copy-paste), but the command is `copyObject`. You'll see the required change in the next version.